Scrape 2016 term

ondenman opened the pull request:
New term data is available on the webpage.

The data types and layout differs from the previous term. 

The member list is presented in a single table. It gives data for name, district and party.

Source: http://www.palemene.ws/new/members-of-parliament/members-of-the-xvi-parliament/
## New data (scraper)
- [x] scraper gets current (i.e. new term's) members (are they _really_ the current members?)
- [x] new term data has correct term set
- [x] fields that are no longer being populated are explained/justified
- [x] new fields that are available are being picked up
- [x] names are in same format as previous term *
- [x] quantities seem sensible (how many politicians _should_ be in this house?)

\* The names are composed as follows: _<tribal names> + <fore and surnames>_ Tribal names are in all caps on the source site. There are differences compared to the live EP data. For example, EP lists the PM as: Tuilaepa Sailele Malielegaoi (this is how wikipedia lists him), though the source site lists him as: 'Auelua Fatialofa Lupesoliai Tuilaepa Dr Sailele Malielegaoi' (His tribal names are: Auelua Fatialofa Lupesoliai Tuilaepa)

All members have tribal names. A person with a tribal name is referred to by their tribal name -- as opposed to their birth name. This article refers to this member by his tribal name, although his birth name is John: http://www.samoaobserver.ws/en/09_03_2016/local/3433/Shy-doctor-gets-his-shot.htm

See:
https://en.wikipedia.org/wiki/Fa%27amatai
http://www.daviddfriedman.com/Academic/Course_Pages/Legal_Systems_Very_Different_13/LegalSysPapers2Discuss13/McHenry_The_Samoan_Way.htm

The source site lists member names prefixed with formal terms of address: afioga, tōfā, susuga. These have not been captured. (https://mike.bitrevision.com/samoan/word/search/?s=Susuga)

ondenman committed:
Change source URL

Updates URL for source of 2016 term members.

ondenman committed:
Update scraper for 2016 term page

The member list is presented in a single table. It gives data
for name, district and party.

Source: http://www.palemene.ws/new/members-of-parliament/members-of-the-xvi-parliament/

tmtmtmtm commented on scraper.rb:
```diff
@@ -18,21 +18,25 @@ def noko_for(url)
 end
 
+def name_and_id(str)
+end
```
What is this function for?

tmtmtmtm commented on scraper.rb:
```diff
@@ -18,21 +18,25 @@ def noko_for(url)
-      name: a.attr('data-title'),
-      image: a.attr('data-src'),
+      name: name_and_party.split('(')[0].tidy,
+      party: name_and_party.match(/\(([A-Z]+)/)[1].tidy,
```
This is the `party_id`. You also need to add the `party` as well.

tmtmtmtm commented on scraper.rb:
```diff
@@ -18,21 +18,25 @@ def noko_for(url)
+    ScraperWiki.save_sqlite([:name], data)
   end
 
   unless (next_page = noko.css('div.ngg-navigation a.next/@href')).empty?
```
Is this section still required?

tmtmtmtm reviewed (changes requested):
You should include the relevant section of the checklist here, as it would reveal several key steps that haven't been done…

ondenman committed:
Remove superfluous function

ondenman committed:
Remove condition that checks for link to next page

This is no longer needed as the list is presented on a single page.

ondenman committed:
Record party names for each member

ondenman committed:
Use #split to separate party id from member name

Previously, we were using a regex but that failed to capture the
party name of one member. The regex was looking for uppercase
letters that were preceded by a '(' but the member had two sets
of letters wrapped in parentheses.

The party_id is now separated from the member name using #split.
The last element is used as the party_id is
always the last set of letters wrapped in brackets.

ondenman committed:
Add term to data

ondenman committed:
Extract logic from scraper.rb into Members class

Member names will require some additional parsing. Making the
scraper class based in preparation of that.

ondenman committed:
Update Ruby version to 2.3.1

ondenman committed:
Add field_serializer gem

ondenman committed:
Add field_serializer to Members

The next step will be to extract the logic for scraping
individual members to a separate class. In this commit, this logic
has been moved from #to_a to a temporary home in #legislative_members.

ondenman committed:
Extract logic for scraping members into Member class

Member class represents an individual member of the legislature.

ondenman committed:
Add #full_name

Method returns enitre name, tribal titles and all.

Method created as a step towards separating member names into
constituent parts.

ondenman committed:
Separate Matai names

Separates Matai names from full name and title, returning them
in an array.

ondenman committed:
Add :honorific_prefix field

Stores all matai names as a single string.

ondenman committed:
Separate matai names with semicolon

So that they can be handle downstream as separate values.

ondenman committed:
Separate name from titles

#name_without_titles splits the fore and surnames from the formal address and tribal titles.

ondenman committed:
Initialize Memberships with noko document

Instead of passing in a URL, Membership now initializes with
a noko document.

This will be useful in a future commit where the document will
be altered to normalize the names before scraping.

ondenman committed:
Capitalize tribal names of Prime Minister

All members except the Prime Minister have their trible
names listed in all-caps. This causes complications
in scraping as the trible names are differentiated by
their capitalization. To avoid complications we convert
the trible names to uppercase before passing the document
to Memberships.

ondenman committed:
Combine fore and surnames with matai titles

This article refers to the member by her matai title: http://www.samoaobserver.ws/en/18_03_2016/local/3805/Fiame-named-Deputy-Prime-Minister-Fonotoe-Faumuina-out-of-Cabinet.htm

From this, it's clear are considered to be actual names rather than prefixes.

This commit combines the fore and surnames of the members with their matai
titles and returns the combined string in the `:name` field.

ondenman committed:
Add :other_name field

Stores name without tribal titles (ie: Fore and surnames)

ondenman committed:
Remove :honorific_prefix field

Matai titles seem to be more than honorifics and a person with
a tribal chief titles is know by that title.

See: http://www.samoaobserver.ws/en/09_03_2016/local/3433/Shy-doctor-gets-his-shot.htm

Dr John Ah Ching is a non native. His title was conferred upon him after
his decision to go into politics. The article refers to him by his
tribal name "Salausa".

I am removing this field as I believe that the trible titles, together with fore
and surnames form the full name of a member.

ondenman committed:
Handle professional titles

Removes professional titles from :name and stores them in :honorifix_prefix

ondenman committed:
Remove :other_name

Removing as I do not have enough information about whether
it is correct to refer to the fore and surnames of a Samoan
tribal chief as an 'other name'.

tmtmtmtm commented on scraper.rb:
```diff
@@ -2,37 +2,23 @@
-  end
+URL = 'http://www.palemene.ws/new/members-of-parliament/members-of-the-xvi-parliament/'.freeze
+# We need to tweak the document slightly as the trible names of one member are
+# not capitalized in the same way as the trible names of other members. We're doing this now
```
typo here, and on line above: trible → tribal

tmtmtmtm commented on lib/members.rb:
```diff
@@ -0,0 +1,28 @@
+    noko.xpath('//div[@class="entry"]/table[1]/tbody/tr').map do |row|
+      Member.new(row).to_h
+    end
+  end
```
It's not obvious why you have a field here that is simply calling another public method.

tmtmtmtm reviewed (changes requested):
The logic for handling all the titles here is so complex that I think this could usefully have a few test-cases.

I also think it's worth switching this to the new `scraped-page` approach now, rather than operating slightly differently from how that works.

ondenman commented:
Closing in favour of: https://github.com/everypolitician-scrapers/samoa-parliament/pull/1
